Make the JXL Makefile work with both tags and SHAs.

veluca93 opened the pull request:
This allows to update the codec without relying on upstream tags.

I didn't make the checkout rule depend on the Makefile, which means that version updates will only come into effect after removing `node_modules`, but it should be possible in principle.

veluca93 committed:
Make the JXL Makefile work with both tags and SHAs.

deymo commented on codecs/jxl/Makefile:
```diff
@@ -78,9 +78,14 @@ $(CODEC_MT_SIMD_BUILD_DIR)/Makefile: CXXFLAGS+=-msimd128
+	rm -rf $(@D)
 	mkdir -p $(@D)
-	git clone $(CODEC_URL) --recursive -j`nproc` --depth 1 --branch $(CODEC_VERSION) $(@D)
+	cd $(@D) && git init .
```
`git init $(@D)`

and then you can do `git -C $(@D) remote add ......` so you don't need to do `cd`

deymo commented on codecs/jxl/Makefile:
```diff
@@ -78,9 +78,14 @@ $(CODEC_MT_SIMD_BUILD_DIR)/Makefile: CXXFLAGS+=-msimd128
+	cd $(@D) && git init .
+	cd $(@D) && git remote add origin $(CODEC_URL)
+	cd $(@D) && git fetch origin $(CODEC_VERSION) --depth 1
+	cd $(@D) && git checkout $(CODEC_VERSION)
```
I'd checkout `FETCH_HEAD` here since the "$(CODEC_VERSION)" may resolve to something different here. For example, if CODEC_VERSION is "master" this will resolve first to the local "master" branch if one exists rather than origin/master.

surma commented:
I actually think we used to have this and we changed it for some reason. @RReverser do you remember?

Either way, this LGTM. Gonna wait for Ingvar to make the final call, tho.

veluca93 commented:
> I actually think we used to have this and we changed it for some reason. @RReverser do you remember?
> 
> Either way, this LGTM. Gonna wait for Ingvar to make the final call, tho.

I think the reason was that the old way didn't do shallow checkouts and/or parallel cloning of submodules (don't remember which, possibly both), but this implementation ought to do both.

RReverser commented:
> I actually think we used to have this and we changed it for some reason. @RReverser do you remember?

Yeah, I added it on my JPEG-XL SIMD branch but then reverted before merging the PR because a new version was tagged, and it didn't seem worth the Makefile complication anymore.

Do we think it's still / again worth it? If there's a usecase (e.g. tags are not catching up fast enough), I don't mind.

> I think the reason was that the old way didn't do shallow checkouts and/or parallel cloning of submodules (don't remember which, possibly both), but this implementation ought to do both.

The one we're talking about, did - it was pretty similar to your PR (I think the only difference is `-C $(@D)` vs `cd ... && git ... && git ...`).

RReverser commented on codecs/jxl/Makefile:
```diff
@@ -78,9 +78,13 @@ $(CODEC_MT_SIMD_BUILD_DIR)/Makefile: CXXFLAGS+=-msimd128
+$(CODEC_DIR)/CMakeLists.txt: 
+	rm -rf $(@D)
+	git init $(@D)
+	git -C $(@D) remote add origin $(CODEC_URL)
```
This line is unnecessary, you can just `fetch $(CODEC_URL) $(CODEC_VERSION)` below.

veluca93 replied:
Nice!

veluca93 commented:
> > I actually think we used to have this and we changed it for some reason. @RReverser do you remember?
> 
> Yeah, I added it on my JPEG-XL SIMD branch but then reverted before merging the PR because a new version was tagged, and it didn't seem worth the Makefile complication anymore.
> 
> Do we think it's still / again worth it? If there's a usecase (e.g. tags are not catching up fast enough), I don't mind.
> 
> > I think the reason was that the old way didn't do shallow checkouts and/or parallel cloning of submodules (don't remember which, possibly both), but this implementation ought to do both.
> 
> The one we're talking about, did - it was pretty similar to your PR (I think the only difference is `-C $(@D)` vs `cd ... && git ... && git ...`).

Yup, it's easier to get bugfixes by checking out a sha instead of waiting for a release on our side (which is actually slightly more involved than just tagging a commit)

RReverser commented on codecs/jxl/Makefile:
```diff
@@ -78,9 +78,12 @@ $(CODEC_MT_SIMD_BUILD_DIR)/Makefile: CXXFLAGS+=-msimd128
-	mkdir -p $(@D)
-	git clone $(CODEC_URL) --recursive -j`nproc` --depth 1 --branch $(CODEC_VERSION) $(@D)
+$(CODEC_DIR)/CMakeLists.txt: 
+	rm -rf $(@D)
```
```suggestion
	$(RM) $(@D)
```

veluca93 replied:
This still needs a -r, right? Done.

RReverser commented on codecs/jxl/Makefile:
```diff
@@ -78,9 +78,12 @@ $(CODEC_MT_SIMD_BUILD_DIR)/Makefile: CXXFLAGS+=-msimd128
+	git init $(@D)
+	git -C $(@D) fetch $(CODEC_URL) $(CODEC_VERSION) --depth 1
+	git -C $(@D) checkout FETCH_HEAD
+	git -C $(@D) submodule update --init --recursive --jobs `nproc`
```
Found one more optimisation that was in my fork but not here:

```suggestion
	git -C $(@D) submodule update --init --depth 1 --recursive --jobs `nproc`
```
